### PR TITLE
Migrate tvb-multiscale to tvb-library v 2.0 + py3

### PR DESCRIPTION
**Pull request by gabiflorea10**
Requirements defined on these tasks:
https://req.thevirtualbrain.org/browse/HBP-36
https://req.thevirtualbrain.org/browse/HBP-37

### CODE REVIEW

File: tvb_scripts/timeseries/model.py
Comment (liadomide) [inline tvb_scripts/timeseries/model.py]: Why is this line removed?

File: tvb_nest/simulator_tvb/simulator.py
Comment (liadomide) [inline tvb_nest/simulator_tvb/simulator.py:84]: required=True has been missed here

File: docker/launch_example.py
Comment (liadomide) [inline docker/launch_example.py]: Pls do not update launch_example. yet. It is in direct relation with HBP Collab usages, and the release Docker image on Docker Hub. We will need to upgrade all of those in correlation with this file. Currently it is fine to remain outdated, this file

File: tvb_nest/interfaces/base.py
Comment (liadomide) [inline tvb_nest/interfaces/base.py]: Why this change ?

Comment (dionperd) [reply tvb_nest/interfaces/base.py]: Indeed. Only the regions that are modelled as NEST nodes should update some of their parameters from NEST. The  rest should keep their user-defined values. 
This mechanism is used when the NEST nodes are also integrated to some extend as TVB nodes, but with some parameter changing at each time point, depending on the simulation in NEST.
In the example I have implemented, an instant rate is computed at each time step by measuring the spikes number per time step of the spiking neurons of each NEST node population. 
Then only for those nodes, this rate value enters as a parameter in the integration of those nodes in TVB.
For all other nodes in TVB the rate parameter is computed from currents, via the specific equations of the model.
I am just giving some background to understand why this change makes no sense.
Also, I don't understand why param_values has changed to param_vals only for that line.

**Comment by dionperd** [comment]
A general question: I guess we will not merge before tvb-library is fully updated by merging with neotraits_py3. Am I write?

**Comment by liadomide** [comment]
> A general question: I guess we will not merge before tvb-library is fully updated by merging with neotraits_py3. Am I right?

The optimal answer would be Yes. But these are still exceptional situations to consider (e.g. how long it takes for neotraits to be ready in the framework. tvb-library should be stable enough already in that branch)

File: tvb_nest/simulator_tvb/simulator.py
Comment (gabiflorea10) [reply tvb_nest/simulator_tvb/simulator.py:84]: Required=True has been removed because List from _attr.py (that defines monitors) hasn't the attribute 'required' as parameter. Yet, because of the way this class is written, the attribute required=True is set internally in List.

File: tvb_scripts/timeseries/model.py
Comment (gabiflorea10) [reply tvb_scripts/timeseries/model.py]: This line has been removed only for passing over some code, without throwing errors. The errors have been solved and the correct code can be seen in a future commit.

File: docker/launch_example.py
Comment (gabiflorea10) [reply docker/launch_example.py]: Ok. This file will be reverted.

File: tvb_nest/interfaces/base.py
Comment (gabiflorea10) [reply tvb_nest/interfaces/base.py]: param_values is a numpy ndarray. It is immutable.. so we cannot assign a value to it (it is a read-only attribute). However, if we want, we can force this ndarray to become mutable, by setting the writable flag to True (something like: 'name.setflags(write=True)'). But after we solve this problem, there is another one: param_values looks like this: [ 1. ] … But self.nest_nodes_ids looks like: [34 35]. So we try to set some values in an ndarray at some out of bound locations. I don't understand exactly why this happens (only when running with library 2.0), but I observed (at first sight) that is not used in critical places in project, so I have decided to use for the moment a new variable for this purpose.

**Review by dionperd — COMMENTED** [review]
Whether the TVB models' parameters should be read-only or not, is something you guys have to decide, taking into consideration the fact that there might be an implementation like this one in tvb-multiscale, where NEST needs to update those parameters only for selected nodes.

Also, I don't understand how this can be read only when the user can at any moment modify these values manually, e.g., by running something like:
simulator.model.a = 1.0 or whatever else.

Could we, anyway, configure these traits with name.setflags(write=True) only for tvb-multiscale?

As for "param_values looks like this: [ 1. ]" 
This is not true, because during the configuration of the simulator in tvb-multiscale I make sure that they all have a length equal to the number of regions (lines 234-238 in tvb-multiscale/tvb_nest/simulator_tvb/simulator.py):
dummy = numpy.ones((self.connectivity.number_of_regions, 1))
# Confirm good shape for NEST to TVB model parameters
for param in self.tvb_nest_interface.nest_to_tvb_params:
        setattr(self.model, "__" + param,
       (dummy * numpy.array(getattr(self.model, param))).squeeze())

Anyway, there is no point in creating param_vals and never be used in the following lines! In the end, it has to modify the model's parameters one way or another for this specific way of NEST -> TVB communication to take place.

If you wish, we can talk (e.g., Skype), so that I understand more about the trait system, and you about the logic of the code here.

File: tvb_nest/simulator_tvb/simulator.py
Comment (liadomide) [reply tvb_nest/simulator_tvb/simulator.py:84]: Thank you for the explanation. Then your change makes sense.

File: tvb_scripts/timeseries/model.py
Comment (liadomide) [inline tvb_scripts/timeseries/model.py:256]: Good catch! The change is ok

File: tvb_nest/simulator_tvb/simulator.py
Comment (liadomide) [inline tvb_nest/simulator_tvb/simulator.py:196]: Well found spot. This is critical after the neotraits switch!

**Review by liadomide — CHANGES_REQUESTED** [review]
Current changes look decent to me.
It remain to have the PY example/test running, and to update Dockerfile to Python 3

File: tvb_nest/tests/test_launch_default_simulation.py
Comment (liadomide) [inline tvb_nest/tests/test_launch_default_simulation.py]: There seem be no actual need to have these stored into files. Stay to the return of result in memory

File: tvb_nest/tests/test_launch_default_simulation.py
Comment (liadomide) [inline tvb_nest/tests/test_launch_default_simulation.py:139]: Maybe also check that the result has numbers, and does not contain Inf or NaN

File: tvb_nest/tests/test_reduced_wong_wang_exc_io_inh_i.py
Comment (liadomide) [inline tvb_nest/tests/test_reduced_wong_wang_exc_io_inh_i.py:4]: After discussing and understanding the previous unit-test we decided it is best to keep the octave dependency, enhance the dockerimage to include octave and oct2py python library and run the original test

**Review by liadomide — CHANGES_REQUESTED** [review]
Good direction. Some changes discussed for the unit-tests

File: tvb_nest/tests/test_reduced_wong_wang_exc_io_inh_i.py
Comment (gabiflorea10) [reply tvb_nest/tests/test_reduced_wong_wang_exc_io_inh_i.py:4]: Ok. The suggestion has been applied and it can be seen in a future commit. In this case, the file rww_dfun.py, added at the latest commit, should be ignored at the merge.

**Pull request merged**
